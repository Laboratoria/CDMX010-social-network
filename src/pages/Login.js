export default ({ signIn }) => {
  const render = () => `
    <div>
      <form>
        <label for="email">Email</label></br>
        <input type"email" id="email" name="email" value=""></br>
        <label for="password">Password</label></br>
        <input type"password" name="password" id="password" value=""></br>
        <input type="submit" id="submitBtn" value="Sign In">
      </form>
      <p id="errorMessage"></p>
    </div>    
  `;

  const attachListeners = () => {
    const submitBtn = document.getElementById('submitBtn');

    submitBtn.addEventListener('click', (e) => {
      e.preventDefault();
      const email = document.getElementById('email').value;
      const password = document.getElementById('password').value;
      signIn(email, password)
        .catch((error) => {
          document.getElementById('errorMessage').innerHTML = error.message;
        });
    });
  };

  return { render, attachListeners };
};
