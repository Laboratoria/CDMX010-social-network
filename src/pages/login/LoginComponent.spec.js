/* eslint-disable prefer-promise-reject-errors */
/* eslint-disable no-unused-vars */
import { screen, waitFor } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import Login from './LoginComponent';

describe('LoginComponent', () => {
  beforeEach(() => {
    document.body.innerHTML = '<div id="root"></div>';
  });

  it('should render', () => {
    const target = document.getElementById('root');
    Login({ firebaseClient: {} }).render(target);

    expect(target.innerHTML).toMatchSnapshot();
  });

  it('should login the user when the user submits the login form', () => {
    const target = document.getElementById('root');
    const signIn = jest.fn().mockImplementation(() => Promise.resolve('ok'));
    const email = 'test@example.com';
    const password = '123456789';
    Login({ signIn }).render(target);

    userEvent.type(screen.getByLabelText('Email'), email);
    userEvent.type(screen.getByLabelText('Password'), password);
    userEvent.click(screen.getByDisplayValue('Sign In'));
    expect(signIn).toHaveBeenCalledWith(email, password);
  });

  it('should show an error message when a user provides invalid credentials', async () => {
    const target = document.getElementById('root');
    const message = 'Invalid credentials!';
    const signIn = jest.fn().mockImplementation(
      (email, password) => Promise.reject({ message }),
    );
    const email = 'test@example.com';
    const password = 'invalidpassword';
    Login({ signIn }).render(target);

    userEvent.type(screen.getByLabelText('Email'), email);
    userEvent.type(screen.getByLabelText('Password'), password);
    userEvent.click(screen.getByDisplayValue('Sign In'));
    expect(signIn).toHaveBeenCalledWith(email, password);

    await waitFor(() => {
      expect(screen.getByTestId('errorMessage').innerHTML).toBe(message);
    });
  });
});
