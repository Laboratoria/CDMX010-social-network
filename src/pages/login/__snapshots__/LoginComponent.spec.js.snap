// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent should render 1`] = `
"
      <div>
        <form>
          <label for=\\"email\\">Email</label><br>
          <input type\\"email\\"=\\"\\" id=\\"email\\" name=\\"email\\" value=\\"\\"><br>
          <label for=\\"password\\">Password</label><br>
          <input type\\"password\\"=\\"\\" name=\\"password\\" id=\\"password\\" value=\\"\\"><br>
          <input type=\\"submit\\" value=\\"Sign In\\">
        </form>
        <p id=\\"errorMessage\\" data-testid=\\"errorMessage\\"></p>
      </div>    
    "
`;
